### ISSUE TITLE
Support -Zpolymorphize

### ISSUE THREAD

**Issue by bjorn3** [bjorn3/rustc_codegen_cranelift#1133]

**Comment by bjorn3** [comment]
```
error: compilation failed!
status: exit code: 101
command: "/home/bjorn/Projects/cg_clif/rust/../build/bin/cg_clif" "/home/bjorn/Projects/cg_clif/rust/src/test/codegen-units/polymorphization/unused_type_parameters.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "-Z" "human_readable_cgu_names" "-C" "prefer-dynamic" "--out-dir" "/home/bjorn/Projects/cg_clif/rust/build/x86_64-unknown-linux-gnu/test/codegen-units/polymorphization/unused_type_parameters" "-Crpath" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/home/bjorn/Projects/cg_clif/rust/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Zpolymorphize=on" "-Zprint-mono-items=lazy" "-Copt-level=1" "-L" "/home/bjorn/Projects/cg_clif/rust/build/x86_64-unknown-linux-gnu/test/codegen-units/polymorphization/unused_type_parameters/auxiliary"
stdout:
------------------------------------------
MONO_ITEM fn <u32 as std::default::Default>::default @@ unused_type_parameters.3a1fbbbh-cgu.0[Internal] unused_type_parameters.3a1fbbbh-cgu.1[Internal] unused_type_parameters.3a1fbbbh-cgu.2[Internal] unused_type_parameters.3a1fbbbh-cgu.3[Internal]
MONO_ITEM fn <u64 as std::default::Default>::default @@ unused_type_parameters.3a1fbbbh-cgu.0[Internal] unused_type_parameters.3a1fbbbh-cgu.1[Internal] unused_type_parameters.3a1fbbbh-cgu.2[Internal] unused_type_parameters.3a1fbbbh-cgu.3[Internal]
MONO_ITEM fn closures::no_parameters @@ unused_type_parameters.3a1fbbbh-cgu.1[External]
MONO_ITEM fn closures::unused::<T> @@ unused_type_parameters.3a1fbbbh-cgu.1[External]
MONO_ITEM fn closures::unused::<T>::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.1[Internal]
MONO_ITEM fn closures::used_argument::<u32> @@ unused_type_parameters.3a1fbbbh-cgu.1[External]
MONO_ITEM fn closures::used_argument::<u32>::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.1[Internal]
MONO_ITEM fn closures::used_argument::<u64> @@ unused_type_parameters.3a1fbbbh-cgu.1[External]
MONO_ITEM fn closures::used_argument::<u64>::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.1[Internal]
MONO_ITEM fn closures::used_argument_closure::<u32> @@ unused_type_parameters.3a1fbbbh-cgu.1[External]
MONO_ITEM fn closures::used_argument_closure::<u32>::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.1[Internal]
MONO_ITEM fn closures::used_argument_closure::<u64> @@ unused_type_parameters.3a1fbbbh-cgu.1[External]
MONO_ITEM fn closures::used_argument_closure::<u64>::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.1[Internal]
MONO_ITEM fn closures::used_binding_type::<u32> @@ unused_type_parameters.3a1fbbbh-cgu.1[External]
MONO_ITEM fn closures::used_binding_type::<u32>::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.1[Internal]
MONO_ITEM fn closures::used_binding_type::<u64> @@ unused_type_parameters.3a1fbbbh-cgu.1[External]
MONO_ITEM fn closures::used_binding_type::<u64>::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.1[Internal]
MONO_ITEM fn closures::used_binding_value::<u32> @@ unused_type_parameters.3a1fbbbh-cgu.1[External]
MONO_ITEM fn closures::used_binding_value::<u32>::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.1[Internal]
MONO_ITEM fn closures::used_binding_value::<u64> @@ unused_type_parameters.3a1fbbbh-cgu.1[External]
MONO_ITEM fn closures::used_binding_value::<u64>::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.1[Internal]
MONO_ITEM fn closures::used_parent::<T>::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.1[Internal]
MONO_ITEM fn closures::used_parent::<u32> @@ unused_type_parameters.3a1fbbbh-cgu.1[External]
MONO_ITEM fn closures::used_parent::<u64> @@ unused_type_parameters.3a1fbbbh-cgu.1[External]
MONO_ITEM fn closures::used_substs::<u32> @@ unused_type_parameters.3a1fbbbh-cgu.1[External]
MONO_ITEM fn closures::used_substs::<u32>::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.1[Internal]
MONO_ITEM fn closures::used_substs::<u64> @@ unused_type_parameters.3a1fbbbh-cgu.1[External]
MONO_ITEM fn closures::used_substs::<u64>::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.1[Internal]
MONO_ITEM fn closures::used_upvar::<u32> @@ unused_type_parameters.3a1fbbbh-cgu.1[External]
MONO_ITEM fn closures::used_upvar::<u32>::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.1[Internal]
MONO_ITEM fn closures::used_upvar::<u64> @@ unused_type_parameters.3a1fbbbh-cgu.1[External]
MONO_ITEM fn closures::used_upvar::<u64>::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.1[Internal]
MONO_ITEM fn dispatch::<u32> @@ unused_type_parameters.3a1fbbbh-cgu.0[Internal]
MONO_ITEM fn dispatch::<u64> @@ unused_type_parameters.3a1fbbbh-cgu.0[Internal]
MONO_ITEM fn foo @@ unused_type_parameters.3a1fbbbh-cgu.0[External]
MONO_ITEM fn functions::no_parameters @@ unused_type_parameters.3a1fbbbh-cgu.2[External]
MONO_ITEM fn functions::unused::<T> @@ unused_type_parameters.3a1fbbbh-cgu.2[External]
MONO_ITEM fn functions::used_argument::<u32> @@ unused_type_parameters.3a1fbbbh-cgu.2[External]
MONO_ITEM fn functions::used_argument::<u64> @@ unused_type_parameters.3a1fbbbh-cgu.2[External]
MONO_ITEM fn functions::used_binding_type::<u32> @@ unused_type_parameters.3a1fbbbh-cgu.2[External]
MONO_ITEM fn functions::used_binding_type::<u64> @@ unused_type_parameters.3a1fbbbh-cgu.2[External]
MONO_ITEM fn functions::used_binding_value::<u32> @@ unused_type_parameters.3a1fbbbh-cgu.2[External]
MONO_ITEM fn functions::used_binding_value::<u64> @@ unused_type_parameters.3a1fbbbh-cgu.2[External]
MONO_ITEM fn functions::used_substs::<u32> @@ unused_type_parameters.3a1fbbbh-cgu.2[External]
MONO_ITEM fn functions::used_substs::<u64> @@ unused_type_parameters.3a1fbbbh-cgu.2[External]
MONO_ITEM fn methods::Foo::<F>::closure_unused_all::<G> @@ unused_type_parameters.3a1fbbbh-cgu.3[External]
MONO_ITEM fn methods::Foo::<F>::closure_unused_all::<G>::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.3[Internal]
MONO_ITEM fn methods::Foo::<F>::closure_used_fn::<u32> @@ unused_type_parameters.3a1fbbbh-cgu.3[External]
MONO_ITEM fn methods::Foo::<F>::closure_used_fn::<u32>::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.3[Internal]
MONO_ITEM fn methods::Foo::<F>::closure_used_fn::<u64> @@ unused_type_parameters.3a1fbbbh-cgu.3[External]
MONO_ITEM fn methods::Foo::<F>::closure_used_fn::<u64>::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.3[Internal]
MONO_ITEM fn methods::Foo::<F>::unused_both::<G> @@ unused_type_parameters.3a1fbbbh-cgu.3[External]
MONO_ITEM fn methods::Foo::<F>::unused_impl @@ unused_type_parameters.3a1fbbbh-cgu.3[External]
MONO_ITEM fn methods::Foo::<F>::used_fn::<u32> @@ unused_type_parameters.3a1fbbbh-cgu.3[External]
MONO_ITEM fn methods::Foo::<F>::used_fn::<u64> @@ unused_type_parameters.3a1fbbbh-cgu.3[External]
MONO_ITEM fn methods::Foo::<u32>::closure_used_both::<u32> @@ unused_type_parameters.3a1fbbbh-cgu.3[External]
MONO_ITEM fn methods::Foo::<u32>::closure_used_both::<u32>::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.3[Internal]
MONO_ITEM fn methods::Foo::<u32>::closure_used_impl::<G> @@ unused_type_parameters.3a1fbbbh-cgu.3[External]
MONO_ITEM fn methods::Foo::<u32>::closure_used_impl::<G>::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.3[Internal]
MONO_ITEM fn methods::Foo::<u32>::closure_used_substs @@ unused_type_parameters.3a1fbbbh-cgu.3[External]
MONO_ITEM fn methods::Foo::<u32>::closure_used_substs::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.3[Internal]
MONO_ITEM fn methods::Foo::<u32>::used_both::<u32> @@ unused_type_parameters.3a1fbbbh-cgu.3[External]
MONO_ITEM fn methods::Foo::<u32>::used_impl @@ unused_type_parameters.3a1fbbbh-cgu.3[External]
MONO_ITEM fn methods::Foo::<u32>::used_substs @@ unused_type_parameters.3a1fbbbh-cgu.3[External]
MONO_ITEM fn methods::Foo::<u64>::closure_used_both::<u64> @@ unused_type_parameters.3a1fbbbh-cgu.3[External]
MONO_ITEM fn methods::Foo::<u64>::closure_used_both::<u64>::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.3[Internal]
MONO_ITEM fn methods::Foo::<u64>::closure_used_impl::<G> @@ unused_type_parameters.3a1fbbbh-cgu.3[External]
MONO_ITEM fn methods::Foo::<u64>::closure_used_impl::<G>::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.3[Internal]
MONO_ITEM fn methods::Foo::<u64>::closure_used_substs @@ unused_type_parameters.3a1fbbbh-cgu.3[External]
MONO_ITEM fn methods::Foo::<u64>::closure_used_substs::{closure#0} @@ unused_type_parameters.3a1fbbbh-cgu.3[Internal]
MONO_ITEM fn methods::Foo::<u64>::used_both::<u64> @@ unused_type_parameters.3a1fbbbh-cgu.3[External]
MONO_ITEM fn methods::Foo::<u64>::used_impl @@ unused_type_parameters.3a1fbbbh-cgu.3[External]
MONO_ITEM fn methods::Foo::<u64>::used_substs @@ unused_type_parameters.3a1fbbbh-cgu.3[External]
Instance { def: Item(WithOptConstParam { did: DefId(0:21 ~ unused_type_parameters[8787]::closures::used_parent), const_param_did: None }), substs: [u32] } _ZN22unused_type_parameters8closures11used_parent17h07203b5afd39c3caE

------------------------------------------
stderr:
------------------------------------------
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `[closure@/home/bjorn/Projects/cg_clif/rust/src/test/codegen-units/polymorphization/unused_type_parameters.rs:73:23: 73:37]`,
 right: `[closure@/home/bjorn/Projects/cg_clif/rust/src/test/codegen-units/polymorphization/unused_type_parameters.rs:73:23: 73:37]`: Can't write value with incompatible type [closure@/home/bjorn/Projects/cg_clif/rust/src/test/codegen-units/polymorphization/unused_type_parameters.rs:73:23: 73:37] to place with type [closure@/home/bjorn/Projects/cg_clif/rust/src/test/codegen-units/polymorphization/unused_type_parameters.rs:73:23: 73:37]

[u32]
[CPlace { inner: Var(_0, Variable(0)), layout: TyAndLayout { ty: u32, layout: Layout { fields: Primitive, variants: Single { index: 0 }, abi: Scalar(Scalar { value: Int(I32, false), valid_range: 0..=4294967295 }), largest_niche: None, align: AbiAndPrefAlign { abi: Align { pow2: 2 }, pref: Align { pow2: 2 } }, size: Size { raw: 4 } } } }, CPlace { inner: Var(_1, Variable(1)), layout: TyAndLayout { ty: u32, layout: Layout { fields: Primitive, variants: Single { index: 0 }, abi: Scalar(Scalar { value: Int(I32, false), valid_range: 0..=4294967295 }), largest_niche: None, align: AbiAndPrefAlign { abi: Align { pow2: 2 }, pref: Align { pow2: 2 } }, size: Size { raw: 4 } } } }, CPlace { inner: Addr(Pointer { base: Dangling(Align { pow2: 3 }), offset: Offset32(0) }, None), layout: TyAndLayout { ty: [closure@/home/bjorn/Projects/cg_clif/rust/src/test/codegen-units/polymorphization/unused_type_parameters.rs:73:23: 73:37], layout: Layout { fields: Arbitrary { offsets: [], memory_index: [] }, variants: Single { index: 0 }, abi: Aggregate { sized: true }, largest_niche: None, align: AbiAndPrefAlign { abi: Align { pow2: 0 }, pref: Align { pow2: 3 } }, size: Size { raw: 0 } } } }, CPlace { inner: Var(_3, Variable(2)), layout: TyAndLayout { ty: &[closure@/home/bjorn/Projects/cg_clif/rust/src/test/codegen-units/polymorphization/unused_type_parameters.rs:73:23: 73:37], layout: Layout { fields: Primitive, variants: Single { index: 0 }, abi: Scalar(Scalar { value: Pointer, valid_range: 1..=18446744073709551615 }), largest_niche: Some(Niche { offset: Size { raw: 0 }, scalar: Scalar { value: Pointer, valid_range: 1..=18446744073709551615 } }), align: AbiAndPrefAlign { abi: Align { pow2: 3 }, pref: Align { pow2: 3 } }, size: Size { raw: 8 } } } }, CPlace { inner: Addr(Pointer { base: Stack(ss0), offset: Offset32(0) }, None), layout: TyAndLayout { ty: (u32,), layout: Layout { fields: Arbitrary { offsets: [Size { raw: 0 }], memory_index: [0] }, variants: Single { index: 0 }, abi: Scalar(Scalar { value: Int(I32, false), valid_range: 0..=4294967295 }), largest_niche: None, align: AbiAndPrefAlign { abi: Align { pow2: 2 }, pref: Align { pow2: 3 } }, size: Size { raw: 4 } } } }]

function u0:5() -> i32 system_v {
    ss0 = explicit_slot 16
    sig0 = () -> i32 system_v
    fn0 = colocated u0:0 sig0

                                block0:
                                    nop 
                                    jump block1

                                block1:
                                    nop 
@0000                               v0 = call fn0()
@0000                               jump block2

                                block2:
@0000                               nop 
@0001                               jump block3

                                block3:
@0001                               nop 
@0003                               v1 = iconst.i64 8
@0004                               v2 = iconst.i32 3
@0004                               stack_store v2, ss0
}

', src/abi/pass_mode.rs:265:5
stack backtrace:
   0: rust_begin_unwind
             at /rustc/0fc6756b42e0556cc2e18079f5fc6b4d58f4e81a/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/0fc6756b42e0556cc2e18079f5fc6b4d58f4e81a/library/core/src/panicking.rs:92:14
   2: rustc_codegen_cranelift::value_and_place::assert_assignable
   3: rustc_codegen_cranelift::abi::pass_mode::adjust_arg_for_abi
   4: rustc_codegen_cranelift::abi::codegen_terminator_call
   5: rustc_codegen_cranelift::base::codegen_fn_content
   6: rustc_session::utils::<impl rustc_session::session::Session>::time
   7: rustc_codegen_cranelift::base::codegen_fn
   8: rustc_codegen_cranelift::driver::aot::module_codegen
   9: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task
  10: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
  11: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
  12: rustc_codegen_cranelift::driver::time
  13: rustc_codegen_cranelift::driver::aot::run_aot
  14: rustc_codegen_cranelift::driver::codegen_crate
  15: <rustc_codegen_cranelift::CraneliftCodegenBackend as rustc_codegen_ssa::traits::backend::CodegenBackend>::codegen_crate
  16: rustc_interface::passes::QueryContext::enter
  17: rustc_interface::queries::Queries::ongoing_codegen
  18: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  19: rustc_span::with_source_map
  20: rustc_interface::interface::create_compiler_and_run
  21: rustc_span::with_session_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.52.0-nightly (0fc6756b4 2021-02-08) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z threads=1 -Z human_readable_cgu_names -Z unstable-options -Z polymorphize=on -Z print-mono-items=lazy -C prefer-dynamic -C rpath -C debuginfo=0 -C opt-level=1

query stack during panic:
end of query stack

------------------------------------------
```

**Comment by bjorn3** [comment]
Fixed by fd2669d1e91027c4ce31f9a3690ce6dab7d8619a.
